### PR TITLE
All.sh add support for tf-psa-crypto components

### PR DESCRIPTION
**Pull request by mpg**
## Description

- Add `tf-psa-crypto/tests/scripts/all.sh` that can be invoked independently and will source components from `tf-psa-crypto/tests/scripts/component-*.sh`. Ultimately this is going to be the main entry point for tf-psa-crypto's CI when all relevant components have been migrated.
- Turn `tests/scripts/all.sh` into a wrapper that handles both mbedtls and tf-psa-crypto. This is a transitional wapper so that the CI can just keep calling the same single entry point, and still get all components, while we move relevant components to tf-psa-crypto, and until tf-psa-crypto gets its own autonomous CI.

## PR checklist

- [x] **changelog** not required because: internal changes, tests only
- [x] **development PR** here
- [x] **framework PR** not required
- [x] **3.6 PR** provided #9754
- [x] **2.28 PR** not required because: 2.28 is too old for that
- **tests**  not required because: manual testing, see below

## Manual testing

- Once the CI is passing, check the outcome file to confirm we're still running the same number of components.
- Anything else reviewers can think of - time to be creative :)

### CODE REVIEW

File: tests/scripts/all.sh
Comment (ronald-cron-arm) [inline tests/scripts/all.sh:28]: It would be more convenient to support that as eventually we will have for example a `test_cmake_out_of_source` component both in Mbed TLS and TF-PSA-Crypto. I understand the logic below should be modified but it does seem possible, see below.

File: tests/scripts/all.sh
Comment (ronald-cron-arm) [inline tests/scripts/all.sh:42]: Rather global args with a wildcard character or matching COMP_MBEDTLS?

**Review by ronald-cron-arm — COMMENTED** [review]
It looks promising to me. However `./tests/scripts/all.sh test_cmake_tf_psa_crypto_out_of_source` does not seem to work (list of arguments empty for mbedtls-all.sh thus running all the components?). As I said in a comment, it would be more convenient if possible to support components with the same name in Mbed TLS and TF-PSA-Crypto. Thinking about it, we would probably need to change something in outcome.csv to differentiate  the test results of such components.

File: tests/scripts/all.sh
Comment (gilles-peskine-arm) [inline tests/scripts/all.sh:2]: Note that this conflicts with https://github.com/Mbed-TLS/mbedtls/pull/9293, which I hope we can merge soon since all it's missing is one approval on one backport.

**Comment by gilles-peskine-arm** [comment]
> we would probably need to change something in outcome.csv to differentiate the test results of such components.

Presumably `MBEDTLS_TEST_CONFIGURATION` should ensure that the component names are different.

Note that https://github.com/Mbed-TLS/mbedtls/pull/9286 (which we plan to merge this quarter) will make it an error to run use the same value of `MBEDTLS_TEST_CONFIGURATION` twice on the same platform.

File: tests/scripts/all.sh
Comment (gilles-peskine-arm) [reply tests/scripts/all.sh:2]: This also conflicts with https://github.com/Mbed-TLS/mbedtls/pull/9286, but that's not quite ready yet, so I suggest that we go 9293, then 9720, then 9286.

Comment (mpg) [reply tests/scripts/all.sh:2]: Sorry, I'm not seeing the conflict with 9293 - it seems to me that the two PRs are not touching the same files. What am I missing?

File: tf-psa-crypto/tests/scripts/components-build-system.sh
Comment (mpg) [reply tf-psa-crypto/tests/scripts/components-build-system.sh:12]: _Eventually_ yes, but IMO only after the transitional wrapper has been removed.

My view of the plan is as follows:
1. We have an `all.sh` in `tf-psa-crypto` but for now we don't invoke it directly on the CI, only through the transitional wrapper in `mbedtls`. (This PR + https://github.com/Mbed-TLS/TF-PSA-Crypto/issues/6 - minimal requirement for tf-psa-crypto to go live.)
2. We move all relevant components to `tf-psa-crypto` so that it gets the level of coverage we want.
3. We change the CI job on `tf-psa-crypto` to stop checking out `mbedtls` and running its `all.sh` - instead we do the normal thing of checking out `tf-psa-crypto` and running its `all.sh`. (In two steps: (a) add the new job, (b) remove the old one.)
4. We remove the transitional wrapper in `mbedtls` and we're now free to rename `tf-psa-crypto` components as we want. If we followed a consistent naming pattern, this can be as simple as `s/_tf_psa_crypto//` in `components-*.sh`.

As long as we have the transitional wrapper, ie we put `tf-psa-crypto` and `mbedtls` in the same bag, I'd really prefer names to be unique within that bag. To get rid of the naming constraints, rather than increasing the complexity associated with the transitional wrapper, let's work towards a state where we no longer need it.

**Comment by mpg** [comment]
> However `./tests/scripts/all.sh test_cmake_tf_psa_crypto_out_of_source` does not seem to work (list of arguments empty for mbedtls-all.sh thus running all the components?)

Aw, silly me, forgot to test the simple case of a single component. Will fix.

File: tests/scripts/all.sh
Comment (gilles-peskine-arm) [reply tests/scripts/all.sh:2]: My bad, there's no merge conflict with 9293 and I don't think there's a semantic conflict either. There's a potential semantic conflict with https://github.com/Mbed-TLS/mbedtls/pull/9286 since it will error out if a component name is duplicated between mbedtls and tf-psa-crypto.

I ran a merge of something that conflicted but that must have been some different branches or with a non-up-to-date local copy and I can't find exactly what I ran yesterday anymore. I have a conflict checker script and it reports no conflict between the current state of this PR and any other PR opened in the last 6 months. #9293 only conflicts in `.gitignore` with a priority-medium PR.

Comment (mpg) [reply tests/scripts/all.sh:2]: > There's a potential semantic conflict with #9286 since it will error out if a component name is duplicated between mbedtls and tf-psa-crypto.

Well, currently my PR also forbids component names that are duplicated between mbedtls and tf-psa-crypto, so that shouldn't be a problem. As I said earlier, I don't think we should allow components with the same name on both sides until the sides are completely separated.

File: tf-psa-crypto/tests/scripts/components-build-system.sh
Comment (ronald-cron-arm) [reply tf-psa-crypto/tests/scripts/components-build-system.sh:12]: Thanks for the explanation. That looks a good plan.

**Comment by mpg** [comment]
Internal CI fully passed, and OpenCI only failed the armcc components which are a known temporary issue.

**Comment by mpg** [comment]
The component that was moved to tf-psa-crypto is not showing up in `outcomes.csv.xz`. I checked the pipeline log and it ran, so the problem is that the outcome is not recorded. I strongly suspect that's become the path to the outcomes file is relative, so the file will be created in `tf-psa-crypto` where it's not going to be picked up later when all the outcomes files from each component are concatenated.

**Comment by mpg** [comment]
> The component that was moved to tf-psa-crypto is not showing up in `outcomes.csv.xz`. I checked the pipeline log and it ran, so the problem is that the outcome is not recorded. I strongly suspect that's become the path to the outcomes file is relative, so the file will be created in `tf-psa-crypto` where it's not going to be picked up later when all the outcomes files from each component are concatenated.

The last commit seems to have fixed this. I downloaded the outcomes file from the Open CI run (Internal CI has not completed yet) then extracted the list of components with `cut -d';' -f2 outcomes-new.csv | sort -u` and compared to the outcomes files of the nightly run from 23 Oct (when this PR was branched out), and the only difference was the various `component_release` that are expected to be run only in the nightly, so everything's as expected.

File: tf-psa-crypto/tests/scripts/components-build-system.sh
Comment (mpg) [reply tf-psa-crypto/tests/scripts/components-build-system.sh:12]: I've now included a copy of this plan in the wrapper's comments for reference.

File: tests/scripts/all.sh
Comment (Harry-Ramsey) [inline tests/scripts/all.sh:58]: If I attempt to run the script using the following command `./tests/scripts/all.sh test_cmake_tf_psa_crypto_out_of_source` I encounter the following error:
```
./tests/scripts/all.sh: line 58: $3: unbound variable
```
I believe this is incorrect and I should either be greeted with the error or the command executes?

**Review by Harry-Ramsey — CHANGES_REQUESTED** [review]

File: tests/scripts/all.sh
Comment (mpg) [reply tests/scripts/all.sh:58]: Ah right, good point, that's poor error handling. I'll fix it.

**Review by Harry-Ramsey — APPROVED** [review]
Thanks, LGTM.

**Review by eleuzi01 — APPROVED** [review]
LGTM, thanks!

File: tf-psa-crypto/tests/scripts/components-build-system.sh
Comment (gilles-peskine-arm) [inline tf-psa-crypto/tests/scripts/components-build-system.sh:18]: Echoing Harry's concern in https://github.com/Mbed-TLS/mbedtls/issues/9755#issuecomment-2460242562 : we currently prevent `component_test_cmake_out_of_source` from running on Ubuntu 16.04, but now `component_test_cmake_tf_psa_crypto_out_of_source` will run on Ubuntu 16.04. Why is this not a problem? I don't see it discussed in the review here.

The problem is a race condition. Just having a small number of successful CI runs is not enough to gain confidence.

Maybe the cmake file in tf-psa-crypto is better written and is not susceptible to the race condition. That's quite plausible, but do we actually know that?

**Review by gilles-peskine-arm — DISMISSED** [review]
Gatekeeper note: I'm blocking merging because I'm concerned that this may introduce an intermittent CI failure. Feel free to override my change request if my concern is invalid.

File: tf-psa-crypto/tests/scripts/components-build-system.sh
Comment (ronald-cron-arm) [reply tf-psa-crypto/tests/scripts/components-build-system.sh:18]: The component `component_test_cmake_tf_psa_crypto_out_of_source` is not added by this PR. This PR moves it from `tests/scripts/components-build-system.sh` to `tf-psa-crypto/tests/scripts/components-build-system.sh`. The component was added by the PR 9445 without protection against Ubuntu 16.04 (we just missed that point I think) which was merged October the 3rd. Thus it has been around for more that a month now. Enough to gain confidence?

Comment (mpg) [reply tf-psa-crypto/tests/scripts/components-build-system.sh:18]: @gilles-peskine-arm I think your concern is valid in general, but quite unrelated to this PR, as Ronald explained.

> Feel free to override my change request if my concern is invalid.

Since you offered, I'm gonna do just that. I'm also going to start a discussion on slack about the potential issue and things we could do about it.

Comment (gilles-peskine-arm) [reply tf-psa-crypto/tests/scripts/components-build-system.sh:18]: Ok, I'd missed that it was a move and not new. I agree that a month of CI is enough to have confidence that this isn't a problem with the tf-psa-crypto cmake scripts.

**Pull request merged**
